src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs

```diff
@@ -38,67 +39,136 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
-        private (ArrayBuilder<BoundExpression> BuilderPatternExpressions, BoundLocal Result) RewriteToInterpolatedStringHandlerPattern(BoundInterpolatedString node)
+        private (ArrayBuilder<BoundExpression> HandlerPatternExpressions, BoundLocal Result) RewriteToInterpolatedStringHandlerPattern(BoundInterpolatedString node)
         {
             Debug.Assert(node.InterpolationData is { Construction: not null });
             Debug.Assert(node.Parts.All(static p => p is BoundCall or BoundStringInsert { Value: BoundCall }));
             var data = node.InterpolationData.Value;
             var builderTempSymbol = _factory.InterpolatedStringHandlerLocal(data.BuilderType, data.ScopeOfContainingExpression, node.Syntax);
-            var builderTemp = _factory.Local(builderTempSymbol);
+            BoundLocal builderTemp = _factory.Local(builderTempSymbol);
 
             // PROTOTYPE(interp-string): Support dynamic creation
-            // var builder = new BuilderType(baseStringLength, numFormatHoles);
-            Debug.Assert(data.Construction is BoundObjectCreationExpression);
-            var builderConstruction = _factory.AssignmentExpression(builderTemp, (BoundExpression)VisitObjectCreationExpression((BoundObjectCreationExpression)data.Construction));
+            // var handler = new HandlerType(baseStringLength, numFormatHoles, ...InterpolatedStringHandlerArgumentAttribute parameters, <optional> out bool handlerIsValid);
+            var construction = (BoundObjectCreationExpression)data.Construction;
+
+            BoundLocal? appendShouldProceedLocal = null;
+            if (data.HasTrailingHandlerValidityParameter)
+            {
+                Debug.Assert(construction.ArgumentRefKindsOpt[^1] == RefKind.Out);
+
+                BoundInterpolatedStringArgumentPlaceholder trailingParameter = data.ArgumentPlaceholders[^1];
+                TypeSymbol localType = trailingParameter.Type!;
+                Debug.Assert(localType.SpecialType == SpecialType.System_Boolean);
+                var outLocal = _factory.SynthesizedLocal(localType);
+                appendShouldProceedLocal = _factory.Local(outLocal);
+
+                AddPlaceholderReplacement(trailingParameter, appendShouldProceedLocal);
+            }
+
+            var handlerConstructionAssignment = _factory.AssignmentExpression(builderTemp, (BoundExpression)VisitObjectCreationExpression(construction));
 
             var usesBoolReturn = data.UsesBoolReturns;
-            var builderPatternExpressions = ArrayBuilder<BoundExpression>.GetInstance(usesBoolReturn ? 2 : node.Parts.Length);
-            builderPatternExpressions.Add(builderConstruction);
-            BoundExpression? appendBinaryExpression = null;
+            var resultExpressions = ArrayBuilder<BoundExpression>.GetInstance(node.Parts.Length + 1);
             foreach (var currentPart in node.Parts)
             {
                 var appendCall = (BoundCall)currentPart;
```

Comment:
The assertion above (`Debug.Assert(node.Parts.All(static p => p is BoundCall or BoundStringInsert { Value: BoundCall }));` on line 49) probably should be tightened to only expect `BoundCall` parts.